### PR TITLE
17 post history

### LINKED ISSUE #17: POST /api/v1/patients/:patientId/history
POST /api/v1/patients/:patientId/history### Description:
Through this route, the user can add a data log to the patient's history  

#### Acceptance Criteria
- [ ] Validation
- [ ] DB query
- [ ] Response
- [ ] Testing
---

### Request Body :
```js
description :'loremloremloremloremlorem ',
cost:'100',
payment:100,

Validation:
description :string.required()
cost:number.min(0).required()
payment:number.min(0).required()
```
### Test:
The Response in success should include:-

#### Response:
```js
{
statusCode:200,
message:'success'
data: [{new history log}]
}
```
The Response in failure should include:-

#### 1- Validation error

statusCode:409

### PR DESCRIPTION
**Pull request by AlaaShurrab**
fixes #17
creating a route to add a history route to the database 
```js
{
    "description":"some sort fo treatment",
    "price":"200",
    "payment":"200"
}
```

### CODE REVIEW

File: server/controllers/patients/addHistoryLog/index.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  getAppointmentsByIdQuery,
+  updateAppointmentStatusQuery,
+} = require('../../../database/queries');
+
+const addHistoryLog = async (req, res, next) => {
+  try {
+    const { appointmentId } = req.params;
+    const { description, price, payment } = req.body;
+
+    const isValid = await historyLogSchema.isValid({
```
Comment (hassanelnajjar) [inline server/controllers/patients/addHistoryLog/index.js:14]: It's good practise to use historyLogSchema.validate(req.body,{abortEarly:false}) function, use can view the exact error of the input fields and send it back to the client

File: server/controllers/patients/addHistoryLog/index.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    const { rows: data } = await addHistoryLogQuery({
+      appointmentId,
+      description,
+      price,
+      payment,
+    });
+
+    if (!appointmentData.is_done) {
```
Comment (hassanelnajjar) [inline server/controllers/patients/addHistoryLog/index.js:48]: I think we don't here a condition to check is_done, it's always false and you need to change it to true

File: server/controllers/patients/addHistoryLog/index.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      await updateAppointmentStatusQuery({ appointmentId });
+    }
+
+    return res.json({
```
Comment (hassanelnajjar) [inline server/controllers/patients/addHistoryLog/index.js:52]: the status here is 201 - and also please the statusCode inside res.body

File: server/database/queries/getAppointmentsByIdQuery/index.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const connection = require('../../config/connection');
+
+const getAppointmentsByIdQuery = ({ appointmentId }) => {
```
Comment (hassanelnajjar) [inline server/database/queries/getAppointmentsByIdQuery/index.js:3]: You here get only the is_done from the appointment record,
I think you need to change the name of this query to be getAppointmentStatus.
And also you don't need to select the appointment id

File: server/controllers/patients/addHistoryLog/index.js (hunk 2 above)
Comment (AlaaShurrab) [reply server/controllers/patients/addHistoryLog/index.js:48]: it's not always false , so if it's true then we don't need to activate the query

Comment (hassanelnajjar) [reply server/controllers/patients/addHistoryLog/index.js:48]: You here you want to add a log for this appointment and change it's status.
If this appointment is_done is true, that's been this appointment is already added to history,it's not correct to change it here. 

I think if is_done is true - you need to throw error

File: server/database/queries/getAppointmentsByIdQuery/index.js (hunk 4 above)
Comment (AlaaShurrab) [reply server/database/queries/getAppointmentsByIdQuery/index.js]: done

File: server/controllers/patients/addHistoryLog/index.js (hunk 3 above)
Comment (AlaaShurrab) [reply server/controllers/patients/addHistoryLog/index.js]: Done

File: server/controllers/patients/addHistoryLog/index.js (hunk 1 above)
Comment (AlaaShurrab) [reply server/controllers/patients/addHistoryLog/index.js]: Done

**Review by hassanelnajjar — APPROVED** [review]

File: server/controllers/patients/addHistoryLog/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+const { boomify, historyLogSchema } = require('../../../utils');
+
+const {
+  addHistoryLogQuery,
+  getPatientProfileData,
+  // getAppointmentsStatusByIdQuery,
+  // updateAppointmentStatusQuery,
```
Comment (Mu7ammadAbed) [inline server/controllers/patients/addHistoryLog/index.js:7]: Remove these

File: server/routes/patients/index.js [hunk 6]
```diff
@@ -3,11 +3,14 @@ const {
...
   getAllProfileData,
   getPatientByNameOrPhone,
+  addHistoryLog,
 } = require('../../controllers');
 
 router.get('/', getPatients);
 router.get('/search', getPatientByNameOrPhone);
 router.route('/:patientId').get(getAllProfileData).patch();
+router.post('/:patientId/history', addHistoryLog);
+router.delete('/:patientId/appointment');
```
Comment (Mu7ammadAbed) [inline server/routes/patients/index.js:13]: remove this

File: server/test/index.test.js [hunk 7]
```diff
@@ -203,6 +205,40 @@ describe('Server Tests', () => {
...
+    });
+    test('addHistoryLogQuery query should add a history log', async () => {
+      const expected = [
+        {
+          id: 3,
+          patient_id: 12,
+          description: 'some sort fo treatment',
+          price: 200,
+          payment: 200,
+          // log_date: new Date('2021-03-31T00:00:00.000Z'),
```
Comment (Mu7ammadAbed) [inline server/test/index.test.js:231]: remove this line

File: server/test/index.test.js [hunk 8]
```diff
@@ -476,6 +511,51 @@ describe('Server Tests', () => {
...
+        title: 'adding a history log',
+        detail: 'data added Successfully',
+        data: [
+          {
+            id: 3,
+            patient_id: 12,
+            description: 'some sort fo treatment',
+            price: 200,
+            payment: 200,
+            // log_date: '2021-03-30T21:00:00.000Z',
```
Comment (Mu7ammadAbed) [inline server/test/index.test.js:525]: Remove this

File: server/utils/validation/historyLogValidation/index.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+const { object, number, string } = require('yup');
+
+const historyLogSchema = object({
+  patientId: number().min(0).required(),
+  description: string().max(200).required(),
```
Comment (Mu7ammadAbed) [inline server/utils/validation/historyLogValidation/index.js:5]: I think limiting the description with 200 characters only is a bad thing

File: server/controllers/patients/addHistoryLog/index.js (hunk 5 above)
Comment (AlaaShurrab) [reply server/controllers/patients/addHistoryLog/index.js:7]: done

File: server/test/index.test.js (hunk 7 above)
Comment (AlaaShurrab) [reply server/test/index.test.js]: done

File: server/test/index.test.js (hunk 8 above)
Comment (AlaaShurrab) [reply server/test/index.test.js]: done

File: server/utils/validation/historyLogValidation/index.js (hunk 9 above)
Comment (AlaaShurrab) [reply server/utils/validation/historyLogValidation/index.js]: done

File: server/routes/patients/index.js (hunk 6 above)
Comment (AlaaShurrab) [reply server/routes/patients/index.js]: if you mean the delete route then I did

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by hassanelnajjar — APPROVED** [review]

**Review by Alaalser — APPROVED** [review]

**Pull request merged**